### ISSUE TITLE
Monadic Interfaces

### ISSUE THREAD

**Issue by BebeSparkelSparkel** [snoyberg/mono-traversable#229]
I'm not sure if this is a good idea or not but I have been thinking about monadic interfaces and the pure interface can just use `Identity` to use the monadic instance or the `RULES` pragma for the specific type function. [Example](https://github.com/BebeSparkelSparkel/mono-traversable/blob/237cf16249b795573ba2a9d4703477e7fe6d5e8f/mono-traversable/src/Data/MonoTraversable.hs#L995-L1014)

**Comment by L0neGamer** [comment]
I like the concept a lot, since it lets us have more generality. I'd recommend looking at the difference in code gen at a core level to see whether we're taking a performance hit even though we're using Identity

**Comment by L0neGamer** [comment]
We also need to think about how this would break backwards compatibility as well as the fact that many structures wouldn't have monadic implementations like this. 

Would we remove the pure versions from the classes and implement them in terms of the Monadic, or have both and define them in terms of each other? Would this mean eg updateMap wouldn't be defined with updateWithKey? Or do we keep the hierarchies as they are?

**Comment by BebeSparkelSparkel** [comment]
I do not think that we can change the interface but going forward we can require that new classes are the monadic version with pure Identity function wrappers with RULES pragmas or have both the monadic and pure classes.
